x-pack/plugins/infra/public/components/waffle/index.tsx

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { findDOMNode } from 'react-dom';
+import styled from 'styled-components';
+import { AutoSizer } from '../auto_sizer';
+import { GroupOfGroups } from './group_of_groups';
+import { GroupOfNodes } from './group_of_nodes';
+
+import {
+  isWaffleMapGroupWithGroups,
+  isWaffleMapGroupWithNodes,
+} from '../../containers/libs/type_guards';
+import { InfraWaffleData, InfraWaffleMapGroup, InfraWaffleOptions } from '../../lib/lib';
+import { applyWaffleMapLayout } from './lib/apply_wafflemap_layout';
+import { Scroll } from './scroll';
+
+interface WafleMapProps {
+  options: InfraWaffleOptions;
+  map: InfraWaffleData;
+}
+
+const initialState = {
+  hasScroll: false,
+};
+
+type WaffleMapState = Readonly<typeof initialState>;
+
+export class Waffle extends React.Component<WafleMapProps, WaffleMapState> {
+  public readonly state: WaffleMapState = initialState;
+  private outer: any;
+  private inner: any;
+  private scrollBarTimer: any;
+
+  public render() {
+    return (
+      <AutoSizer bounds>
+        {({ measureRef, bounds: { width = 0, height = 0 } }) => {
+          const { hasScroll } = this.state;
+          const { map } = this.props;
+          const groupsWithLayout = applyWaffleMapLayout(map, width, height);
+          return (
+            <Container innerRef={(el: any) => measureRef(el)}>
+              <Scroll hasScroll={hasScroll} innerRef={(el: any) => (this.outer = el)}>
+                <MapContainer innerRef={(el: any) => (this.inner = el)}>
+                  {groupsWithLayout.map(this.renderGroup)}
+                </MapContainer>
+              </Scroll>
+            </Container>
```

Comment:
The way these are nested and styled leads to constant recalculations of the size. How about instead we simplify this to

```typescript
            <Container innerRef={(el: any) => measureRef(el)}>
              <MapContainer>
                {groupsWithLayout.map(this.renderGroup)}
              </MapContainer>
            </Container>
```

with the styled components being defined as

```typescript
const Container = styled.div`
  flex: 1 0 0;
  display: flex;
  justify-content: center;
  flex-direction: column;
  overflow-x: hidden;
  overflow-y: auto;
`;

const MapContainer = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  align-content: flex-start;
  padding: 10px;
`;
```

It looks fine to me and performs better - but I might overlooking something.

Reply:
I think what I proposed above does that. for content that is bigger than the scroll area, there shouldn't be a visible difference between vertical centering and aligning to the top, right?